clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -1432,4 +1425,21 @@ public static String[] enumOptions(Class<? extends Enum<?>> enumClass) {
                 .toArray(String[]::new);
     }
 
+    /**
+     * Creates a preallocated file
+     * @param path File path
+     * @param size The size used for pre allocate file, for example 512 * 1025 *1024
+     */
+    public static FileChannel createPreallocatedFile(Path path, int size) throws IOException {
```

Comment:
My suggestion:
1. Create another function preallocatedFile() which will not create a new file but instead it will just preallocate the file i.e. it won't have `StandardOpenOption.CREATE_NEW`
2. Refactor the code such that there is no code duplication between `createPreallocatedFile` and  `preallocatedFile`.
3. You can now use this new `preallocatedFile` function to replace AbstractIndex.scala, lines 121-126 and lines 195-200.